src/librustc_const_eval/eval.rs

```diff
@@ -1145,10 +1152,17 @@ fn lit_to_const<'tcx>(lit: &ast::LitKind,
             infer(Infer(n), tcx, &ty::TyUint(ity), span).map(Integral)
         },
 
-        LitKind::Float(ref n, _) |
+        LitKind::Float(ref n, fty) => {
+            if let Ok(x) = n.parse::<f64>() {
```

Comment:
I don't know why anyone would *want* to do that, but `ConstVal` doesn't have separate variants for f64 and f32 so avoiding the double rounding doesn't seem possible without semi-large changes all throughout const eval. I've been meaning to find a short example program that demonstrates the double-rounding and file an issue. In any case, fixing this existing brokenness seems orthogonal to and out of scope for this PR.